### DISCUSSION TITLE
How do you handle mouse events

### DISCUSSION THREAD

**Discussion by umariani** [vlang/v#13162] [General]
How do you handle mouse events:
 - Mouse Button -- up, down, hold, release
 - Mouse position x-y
 - Mouse movement

**Comment by Wajinn** [comment]
There was a V automation project called Vrobot, (https://github.com/eioo/vrobot).  Unfortunately, it appears to being using the older version of V (0.1), so no longer works.  People will probably need to fork it and update the code.  I think its a very good idea for V to have an automation project, similar to RobotGo, PyAutoGUI, AutoIt, etc...

**Comment by ghost** [comment]
There are three modules that operate the GUI.
`sokol`, `gg`, and `ui`.

Since V is still under development and will continue to evolve, so not well documented yet, the fastest way to learn is to _read and run through all the examples_ by the core V developers.

If you want to use `gg`, you might want to look at this sample

https://github.com/vlang/v/blob/57563599a710d07a3a1978a1fb24e78fc4beda68/examples/2048/2048.v#L814

`sokol`

https://github.com/vlang/v/blob/57563599a710d07a3a1978a1fb24e78fc4beda68/examples/sokol/particles/particles.v#L120


Hmm. .`ui` doesn't have a good example.
But `ui` is just a wrapper for `gg` (and `gg` for `sokol`).
You can use it in the same way as `gg`.

https://github.com/vlang/ui/blob/master/ui.v
And you can find all the `struct`s and `enum`s.

And you will see a few duplicate things in `gg` and `ui`, but these are issues that should be cleaned up in the future. It is just incomplete, so there is no need to be confused about it.

And I highly recommend to read this source as well.
https://github.com/vlang/ui/blob/master/window.v
Once you have a grasp of this, it will be relatively easy to understand how `ui` works.

As Alex, the author, says, the V standard library is surprisingly easy to read, even for a beginner.
Again, I strongly recommend that you read through all the examples, and the standard library as well, even if only what you need. In V, the source code is like documentation in itself.

**Reply by Wajinn** [reply]
Oops, read too quickly and thought you were talking about automation.  But good that kasha gave answers.
